File: problems/297/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
private:
    string serializeHelper(TreeNode *root) {
        if (not root) {
            return "#,";
        }
        
        return to_string(root->val) + "," + serializeHelper(root->left) + serializeHelper(root->right);
    }
    
    TreeNode* deserialize(queue<string>& q) {
        if (empty(q)) {
            return nullptr;
        }
        
        if (q.front() == "#") {
            q.pop();
            
            return nullptr;
        }
        
        TreeNode *node = new TreeNode(stoi(q.front()));
        q.pop();
        
        node->left = deserialize(q);
        node->right = deserialize(q);
        
        return node;
    }
    
    TreeNode *deserializeHelper(string data) {
        queue<string> q;
        istringstream iss(data);
        string token;
        
        while (getline(iss, token, ',')) {
            q.push(token);
        }
        
        return deserialize(q);
    }
    
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        return serializeHelper(root);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        return deserializeHelper(data);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: problems/297/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def helper(node):
            if not node:
                return '#,'
            
            return str(node.val) + ',' + helper(node.left) + helper(node.right)

        return helper(root)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        q = data.split(',')
        q.reverse()
        
        def helper():
            if not q:
                return None
            
            if q[-1] == '#':
                q.pop()
                return None
            
            node = TreeNode(int(q[-1]))
            q.pop()
            
            node.left = helper()
            node.right = helper()
            
            return node
        
        return helper()
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))